README.md
# influx

[![travis build status](https://travis-ci.org/factisresearch/influx.svg?branch=master)](https://travis-ci.org/factisresearch/influx)


ChangeLog.md
# Revision history for influx

## 0.1.0.0  -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/Main.hs
{-# OPTIONS_GHC -F -pgmF htfpp #-}
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Test.Framework
import {-@ HTF_TESTS @-} Database.Influx.API.Tests
import {-@ HTF_TESTS @-} Database.Influx.Internal.Helpers.Tests

main :: IO ()
main = htfMain htf_importedTests

test/Database/Influx/Tests.hs
{-# OPTIONS_GHC -F -pgmF htfpp #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE OverloadedStrings #-}

module Database.Influx.Tests
   ( htf_thisModulesTests
   ) where

import Database.Influx

import Test.Framework
import Data.Maybe (fromJust)
import qualified Data.Text as T

testConfig :: Config
testConfig =
    Config
    { configCreds = Just creds
    , configServer = "http://localhost:8086"
    , configManager = Nothing
    }
    where
      creds =
          Credentials
          { credsUser = "root"
          , credsPassword = "root"
          }

test_ping :: IO ()
test_ping =
    do res <- ping testConfig
       assertBool $
           maybe False ((>= 1) . T.length . unInfluxVersion) res

showDBs :: IO [T.Text]
showDBs =
    do queryRes <- getQuery testConfig Nothing "SHOW DATABASES"
       pure $ map (\(Cons s ()) -> s) (parsedRows queryRes)

test_createDropDB :: IO ()
test_createDropDB =
    do postQuery testConfig Nothing "CREATE DATABASE integration_test"
       dbs <- showDBs
       postQuery testConfig Nothing "DROP DATABASE integration_test"
       dbsAfter <- showDBs
       assertBool $ "integration_test" `elem` dbs && "integration_test" `notElem` dbsAfter

test_properLineProtocol :: IO ()
test_properLineProtocol =
    let line = "cpu,host=server\\ 01,region=uswest value=1i,msg=\"all systems nominal\""
        repr =
            InfluxData
            { dataMeasurement = "cpu"
            , dataTags = [("host", "server 01"), ("region", "uswest")]
            , dataFields = [("value", Integer 1), ("msg", String "all systems nominal")]
            , dataTimestamp = Nothing
            }
    in assertEqual line (serializeInfluxData repr)


test/Database/Influx/API/Tests.hs
{-# OPTIONS_GHC -F -pgmF htfpp #-}
{-# LANGUAGE OverloadedStrings #-}

module Database.Influx.API.Tests
   ( htf_thisModulesTests
   ) where

import Database.Influx

import Data.Monoid ((<>))
import Test.Framework
import qualified Data.Text as T

testConfig :: Config
testConfig =
    Config
    { configCreds = Just creds
    , configServer = "http://localhost:8086"
    , configManager = Nothing
    }
    where
      creds =
          Credentials
          { credsUser = "root"
          , credsPassword = "root"
          }

test_ping :: IO ()
test_ping =
    do res <- ping testConfig
       assertBool $
           maybe False ((>= 1) . T.length . unInfluxVersion) res

showDBs :: IO [T.Text]
showDBs =
    do queryRes <- getQuery testConfig Nothing "SHOW DATABASES"
       pure $ map (\(Cons s ()) -> s) (parsedRows queryRes)

createDb :: DatabaseName -> IO ()
createDb name =
    let query = Query $ "CREATE DATABASE IF NOT EXISTS " <> name
    in postQuery testConfig Nothing query

dropDb :: DatabaseName -> IO ()
dropDb name =
    let query = Query $ "DROP DATABASE IF EXISTS " <> name
    in postQuery testConfig Nothing query

test_createAndDropDB :: IO ()
test_createAndDropDB =
    do createDb "integration_test"
       dbs <- showDBs
       dropDb "integration_test"
       dbsAfter <- showDBs
       assertBool $
           "integration_test" `elem` dbs &&
           "integration_test" `notElem` dbsAfter

withTestDb :: DatabaseName -> IO () -> IO ()
withTestDb name action =
    do dropDb name
       createDb name
       action
       dropDb name

test_writeWithWrongServerAddress :: IO ()
test_writeWithWrongServerAddress =
    do withTestDb db $
           do res <- write wrongConfig defaultWriteParams db [row]
              let isWriteFailureHttpException =
                      case res of
                        WriteFailed (WriteFailureHttpException _) -> True
                        _ -> False
              assertBool isWriteFailureHttpException
    where
      wrongConfig =
          Config
          { configCreds = Nothing
          , configServer = "https://thisisaserverthatdoesnotexist.co.uk:8086"
          , configManager = Nothing
          }
      db = "foobarblub"
      row = InfluxData "issues" [("source", "client-a")] [("jira-id", Integer 1337)] Nothing

test_writeToNonexistentDb :: IO ()
test_writeToNonexistentDb =
    do res <- write testConfig defaultWriteParams db [row]
       let isDbDoesNotExistError =
               case res of
                 WriteFailed (InfluxDbDoesNotExist _errMsg) -> True
                 _ -> False
       assertBool isDbDoesNotExistError
    where
      db = "thisisadatabasethatdoesnotexist"
      row = InfluxData "issues" [("source", "client-a")] [("jira-id", Integer 1337)] Nothing

test_writeInconsistentData :: IO ()
test_writeInconsistentData =
    withTestDb db $
        do res <- write testConfig defaultWriteParams db rows
           let isBadWriteReqError =
                   case res of
                     WriteFailed (BadInfluxWriteRequest _) -> True
                     _ -> False
           assertBool isBadWriteReqError
    where
      db = "foobarblub"
      rows =
          -- writing this fails because the types for "jira-id" don't match
          [ InfluxData "issues" [("source", "client-a")] [("jira-id", Integer 1337)] Nothing
          , InfluxData "issues" [("source", "client-a")] [("jira-id", String "1337")] Nothing
          ]


src/Database/Influx/Types/Core.hs
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE DeriveFoldable #-}
{-# LANGUAGE DeriveTraversable #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}

module Database.Influx.Types.Core
    ( RetentionPolicy
    , DatabaseName
    , Credentials(..)
    , Config(..)
    , EpochPrecision(..)
    , QueryParams(..)
    , defaultQueryParams
    , InfluxVersion(..)
    , Query(..)
    , Value(..)
    , InfluxPoint(..)
    , InfluxTable(..)
    , InfluxResult(..)
    , InfluxResults(..)
    , ParsedTable(..)
    , TimeStamp(..)
    , InfluxData(..)
    , WriteParams(..)
    , defaultWriteParams
    , QueryFailure(..)
    , QueryResponse(..)
    , WriteFailure(..)
    , WriteResponse(..)
    ) where

import Control.Exception (Exception)
import Data.Aeson ((.:), (.:?))
import Data.String (IsString)
import Data.Text (Text)
import Data.Typeable (Typeable)
import Network.HTTP.Client.Conduit (HttpException, Manager)
import qualified Data.Aeson.Types as A
import qualified Data.Scientific as S
import qualified Data.Vector as V

-- | User credentials
data Credentials
    = Credentials
    { credsUser :: !Text
    , credsPassword :: !Text
    } deriving (Show)

data Config = Config
    { configCreds  :: !(Maybe Credentials)
    , configServer :: !String
    , configManager :: !(Maybe Manager)
    }

type RetentionPolicy = Text

data EpochPrecision
    = Hours
    | Minutes
    | Seconds
    | Milliseconds
    | Microseconds
    | Nanoseconds
    deriving (Eq, Show)

type DatabaseName = Text

data QueryParams
    = QueryParams
    { qp_chunkSize :: !(Maybe Int)
    , qp_epoch :: !(Maybe EpochPrecision)
    , qp_retentionPolicy :: !(Maybe RetentionPolicy)
    , qp_database :: !(Maybe DatabaseName)
    } deriving (Show)

defaultQueryParams :: QueryParams
defaultQueryParams =
    QueryParams
    { qp_chunkSize = Nothing
    , qp_epoch = Nothing
    , qp_retentionPolicy = Nothing
    , qp_database = Nothing
    }

newtype InfluxVersion
    = InfluxVersion
    { unInfluxVersion :: Text
    } deriving Show

newtype Query
    = Query
    { unQuery :: Text
    } deriving (Show, IsString)

data Value
    = String !Text
    | Number !S.Scientific
    | Integer !Integer -- ^ only used for serialization
    | Bool !Bool
    | Null
    deriving (Show, Eq)

instance A.FromJSON Value where
    parseJSON val =
        case val of
          A.String s -> pure (String s)
          A.Number n -> pure (Number n)
          A.Bool b -> pure (Bool b)
          A.Null -> pure Null
          A.Object _ -> fail "didn't expect an object as an influx value"
          A.Array _ -> fail "didn't expect an array as an influx value"

newtype InfluxPoint
    = InfluxPoint
    { influxPointValues :: V.Vector Value
    } deriving (Show, A.FromJSON)

data InfluxTable
    = InfluxTable
    { tableName :: Text
    , tableColumns :: V.Vector Text
    , tableValues :: [InfluxPoint]
    } deriving (Show)

instance A.FromJSON InfluxTable where
    parseJSON =
        A.withObject "InfluxTable" $ \o ->
            do tableName <- o .: "name"
               tableColumns <- o .: "columns"
               tableValues <- o .: "values"
               pure InfluxTable {..}

newtype InfluxResults
    = InfluxResults
    { unInfluxResults :: [InfluxResult]
    } deriving (Show)

instance A.FromJSON InfluxResults where
    parseJSON =
        A.withObject "InfluxResults" $ \o ->
            InfluxResults <$> o .: "results"

data InfluxResult
    = InfluxResult
    { resultError :: Maybe String
    , resultTables :: Maybe [InfluxTable]
    } deriving (Show)

instance A.FromJSON InfluxResult where
    parseJSON =
        A.withObject "InfluxResult" $ \o ->
            do resultError <- o .:? "error"
               resultTables <- o .:? "series"
               pure InfluxResult {..}

data ParsedTable t
    = ParsedTable
    { parsedRows :: [t]
    , pointsThatCouldNotBeParsed :: [InfluxPoint]
    } deriving (Show)

newtype TimeStamp
    = TimeStamp
    { unTimeStamp :: Integer
    } deriving (Show)

data InfluxData
    = InfluxData
    { dataMeasurement :: Text
    , dataTags :: [(Text, Text)]
    , dataFields :: [(Text, Value)]
    , dataTimestamp :: Maybe TimeStamp
    }

data WriteParams
    = WriteParams
    { wp_precision :: !(Maybe EpochPrecision)
    , wp_retentionPolicy :: !(Maybe RetentionPolicy)
    }

defaultWriteParams :: WriteParams
defaultWriteParams =
    WriteParams
    { wp_precision = Nothing
    , wp_retentionPolicy = Nothing
    }

data QueryFailure
    = BadInfluxQueryRequest Text -- ^ Unacceptable request (status code 400). Can occur with a syntactically incorrect query. The returned JSON offers further information.
    | QueryFailureHttpException HttpException -- any other 'HttpException'
    | QueryResponseJsonParseError String
    deriving (Show, Typeable)

instance Exception QueryFailure

data QueryResponse a
    = QueryResult a
    | QueryFailed QueryFailure
    deriving (Show, Functor, Foldable, Traversable)

data WriteFailure
    = BadInfluxWriteRequest Text -- ^ Unacceptable request (status code 400). Can occur with a Line Protocol syntax error or if a user attempts to write values to a field that previously accepted a different value type. The returned JSON offers further information.
    | InfluxDbDoesNotExist Text -- ^ Unacceptable request (status code 404). Can occur if a user attempts to write to a database that does not exist. The returned JSON offers further information.
    | InfluxServerError Text -- ^ Status code 500. The system is overloaded or significantly impaired. Can occur if a user attempts to write to a retention policy that does not exist. The returned JSON offers further information.
    | WriteFailureHttpException HttpException -- ^ any other 'HttpException'
    deriving (Show, Typeable)

instance Exception WriteFailure

data WriteResponse
    = WriteSuccessful
    | WriteFailed WriteFailure
    deriving (Show)


src/Database/Influx/Internal/Helpers.hs
{-# LANGUAGE OverloadedStrings #-}

module Database.Influx.Internal.Helpers
    ( urlAppend
    , credsToQueryString
    , epochToBytestring
    , queryParamsToQueryString
    , serializeInfluxData
    , writeParamsToQueryString
    ) where

import Database.Influx.Types.Core

import Control.Arrow (second)
import Data.Maybe (catMaybes, mapMaybe)
import Data.Monoid ((<>))
import Data.Text (Text)
import qualified Data.ByteString as B
import qualified Data.Text as T
import qualified Data.Text.Encoding as T

urlAppend :: String -> String -> String
urlAppend base path = base' ++ "/" ++ path'
  where base' = if last base == '/' then init base else base
        path' = if head path == '/' then tail path else path
    
credsToQueryString :: Credentials -> [(B.ByteString, Maybe B.ByteString)]
credsToQueryString creds =
    fmap (second Just) $
    [ ("u", T.encodeUtf8 (credsUser creds))
    , ("p", T.encodeUtf8 (credsPassword creds))
    ]

epochToBytestring :: EpochPrecision -> B.ByteString
epochToBytestring epoch =
    case epoch of
      Hours -> "h"
      Minutes -> "m"
      Seconds -> "s"
      Milliseconds -> "ms"
      Microseconds -> "us"
      Nanoseconds -> "ns"

queryParamsToQueryString :: QueryParams -> [(B.ByteString, Maybe B.ByteString)]
queryParamsToQueryString opts =
    fmap (second Just) $
    catMaybes
    [ (,) "chunk_size" . T.encodeUtf8 . T.pack . show <$> qp_chunkSize opts
    , (,) "epoch" . epochToBytestring <$> qp_epoch opts
    , (,) "rp" . T.encodeUtf8 <$> qp_retentionPolicy opts
    , (,) "db" . T.encodeUtf8 <$> qp_database opts
    ]

serializeValue :: Value -> Maybe Text
serializeValue v =
    case v of
      Number n -> Just $ T.pack $ show n
      Integer i -> Just $ T.pack (show i) <> "i"
      String s -> Just $ T.pack $ show s
      Bool b ->
        Just $ if b then "true" else "false"
      Null -> Nothing

serializeInfluxData :: InfluxData -> Text
serializeInfluxData d =
    T.intercalate "," (escape (dataMeasurement d) : map serializeTag (dataTags d)) <> " " <>
    T.intercalate "," (mapMaybe serializeField (dataFields d)) <>
    maybe "" (\t -> " " <> serializeTimeStamp t) (dataTimestamp d)
    where
      serializeTag (k, v) =
          escape k <> "=" <> escape v
      serializeField (k, v) =
          ((escape k <> "=") <>) <$> serializeValue v
      serializeTimeStamp t = T.pack $ show $ unTimeStamp t
      escape = T.replace "," "\\," . T.replace " " "\\ "

writeParamsToQueryString :: WriteParams -> [(B.ByteString, Maybe B.ByteString)]
writeParamsToQueryString opts =
    fmap (second Just) $
    catMaybes
    [ (,) "precision" . epochToBytestring <$> wp_precision opts
    , (,) "rp" . T.encodeUtf8 <$> wp_retentionPolicy opts
    ]

src/Database/Influx/Types/FromInfluxPoint.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE TypeOperators #-}

module Database.Influx.Types.FromInfluxPoint
    ( Result(..)
    , Parser, runParser
    , parseEither
    , FromInfluxValue (..)
    , FromInfluxPoint (..)
    , Cons (..)
    ) where

import Database.Influx.Types.Core

import Control.Monad (ap)
import Control.Monad.Reader.Class (MonadReader(..))
import Data.Text (Text)
import qualified Data.HVect as HV
import qualified Data.Scientific as S
import qualified Data.Text as T
import qualified Data.Vector as V

-- | The result of running a 'Parser'.
data Result a
    = Error String
    | Success a
    deriving (Eq, Show)

instance Functor Result where
  fmap _f (Error err) = Error err
  fmap f (Success x) = Success (f x)

instance Applicative Result where
    pure = return
    (<*>) = ap

instance Monad Result where
    return = Success
    Error err >>= _f = Error err
    Success x >>= f = f x
    fail = Error

resultToEither :: Result x -> Either String x
resultToEither r =
    case r of
      Error err -> Left err
      Success x -> Right x 

newtype Parser a
    = Parser
    { runParser :: Maybe EpochPrecision -> Result a
    }

instance Functor Parser where
    fmap f p = Parser (\ep -> f <$> runParser p ep)

instance Applicative Parser where
    pure = return
    (<*>) = ap

instance Monad Parser where
    return x = Parser (\_ep -> return x)
    p >>= f =
        Parser $ \ep ->
            do x <- runParser p ep
               runParser (f x) ep
    fail err = Parser (\_ep -> fail err)

instance MonadReader (Maybe EpochPrecision) Parser where
    ask = Parser (\ep -> pure ep)
    local f x = Parser (\ep -> runParser x (f ep))

parseEither ::
       (x -> Parser a)
    -> Maybe EpochPrecision
    -> x
    -> Either String a
parseEither p ep x = resultToEither (runParser (p x) ep)

class FromInfluxValue a where
    parseInfluxValue :: Value -> Parser a

instance FromInfluxValue Value where
    parseInfluxValue = pure

instance FromInfluxValue Bool where
    parseInfluxValue val =
        case val of
          Bool b -> pure b
          _ -> fail "expected a bool"

instance FromInfluxValue Text where
    parseInfluxValue val =
        case val of
          String s -> pure s
          _ -> fail "expected a string"

instance FromInfluxValue String where
    parseInfluxValue val =
        case val of
          String s -> pure (T.unpack s)
          _ -> fail "expected a string"

instance FromInfluxValue Integer where
    parseInfluxValue val =
        case val of
          Number s ->
              case S.floatingOrInteger s :: Either Double Integer of
                Left _ -> fail "expected an integer, but got a double"
                Right i -> pure i
          Integer i -> pure i
          _ -> fail "expected an integer"

instance FromInfluxValue Int where
    parseInfluxValue val =
        case val of
          Number s ->
              case S.toBoundedInteger s of
                Nothing -> fail "expected an int, but got a double or an out-of-range integer"
                Just i -> pure i
          Integer i ->
              let intMinBound = toInteger (minBound :: Int)
                  intMaxBound = toInteger (maxBound :: Int)
              in if intMinBound <= i && i <= intMaxBound
                   then pure (fromInteger i)
                   else fail "expected an int, but got an out-of-range integer"
          _ -> fail "expected an integer"

instance FromInfluxValue a => FromInfluxValue (Maybe a) where
    parseInfluxValue val =
        case val of
          Null -> pure Nothing
          _ -> Just <$> parseInfluxValue val

{-
instance FromInfluxValue Time.UTCTime where
    parseInfluxValue val =
        case val of
          String s ->
              case Time.parseTimeM True Time.defaultTimeLocale timestampFormat (T.unpack s) of
                Nothing -> fail "could not parse string as timestamp"
                Just time -> pure time
          _ -> fail "expected a time stamp"
        where
          timestampFormat = "%Y-%m-%dT%H:%M:%SZ"
-}

class FromInfluxPoint a where
    parseInfluxPoint :: InfluxPoint -> Parser a

instance FromInfluxPoint InfluxPoint where
    parseInfluxPoint = pure

data Cons a b = Cons { car :: a, cdr :: b }

instance (FromInfluxValue a, FromInfluxPoint b) =>
    FromInfluxPoint (Cons a b) where
    parseInfluxPoint p =
        let v = influxPointValues p
        in if V.length v >= 1
             then
                 Cons
                     <$> parseInfluxValue (V.head v)
                     <*> parseInfluxPoint (InfluxPoint (V.tail v))
             else fail "expected a non-empty vector"

data Singleton a = Singleton a

instance FromInfluxValue a =>
    FromInfluxPoint (Singleton a) where
    parseInfluxPoint p =
        let v = influxPointValues p
        in if V.length v >= 1
            then Singleton <$> parseInfluxValue (V.head v)
            else fail "expected a non-empty vector"

instance FromInfluxPoint () where
    parseInfluxPoint _p = pure ()

instance (FromInfluxValue a, FromInfluxValue b) => FromInfluxPoint (a, b) where
    parseInfluxPoint p =
        do Cons a (Cons b ()) <- parseInfluxPoint p
           pure (a, b)

instance (FromInfluxValue a, FromInfluxValue b, FromInfluxValue c) => FromInfluxPoint (a, b, c) where
    parseInfluxPoint p =
        do Cons a (Cons b (Cons c ())) <- parseInfluxPoint p
           pure (a, b, c)

instance (FromInfluxValue a, FromInfluxValue b, FromInfluxValue c, FromInfluxValue d) => FromInfluxPoint (a, b, c, d) where
    parseInfluxPoint p =
        do Cons a (Cons b (Cons c (Cons d ()))) <- parseInfluxPoint p
           pure (a, b, c, d)

instance (FromInfluxValue a, FromInfluxValue b, FromInfluxValue c, FromInfluxValue d, FromInfluxValue e) => FromInfluxPoint (a, b, c, d, e) where
    parseInfluxPoint p =
        do Cons a (Cons b (Cons c (Cons d (Cons e ())))) <- parseInfluxPoint p
           pure (a, b, c, d, e)

instance (FromInfluxValue a, FromInfluxValue b, FromInfluxValue c, FromInfluxValue d, FromInfluxValue e, FromInfluxValue f) => FromInfluxPoint (a, b, c, d, e, f) where
    parseInfluxPoint p =
        do Cons a (Cons b (Cons c (Cons d (Cons e (Cons f ()))))) <- parseInfluxPoint p
           pure (a, b, c, d, e, f)

instance (FromInfluxValue a, FromInfluxValue b, FromInfluxValue c, FromInfluxValue d, FromInfluxValue e, FromInfluxValue f, FromInfluxValue g) => FromInfluxPoint (a, b, c, d, e, f, g) where
    parseInfluxPoint p =
        do Cons a (Cons b (Cons c (Cons d (Cons e (Cons f (Cons g ())))))) <- parseInfluxPoint p
           pure (a, b, c, d, e, f, g)

instance (FromInfluxValue a, FromInfluxValue b, FromInfluxValue c, FromInfluxValue d, FromInfluxValue e, FromInfluxValue f, FromInfluxValue g, FromInfluxValue h) => FromInfluxPoint (a, b, c, d, e, f, g, h) where
    parseInfluxPoint p =
        do Cons a (Cons b (Cons c (Cons d (Cons e (Cons f (Cons g (Cons h ()))))))) <- parseInfluxPoint p
           pure (a, b, c, d, e, f, g, h)

instance FromInfluxPoint (HV.HVect '[]) where
    parseInfluxPoint _ = pure HV.HNil

instance (FromInfluxValue t, FromInfluxPoint (HV.HVect ts)) =>
    FromInfluxPoint (HV.HVect (t ': ts)) where
    parseInfluxPoint p =
        do Cons x xs <- parseInfluxPoint p
           pure $ x HV.:&: xs


src/Database/Influx/API.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}

module Database.Influx.API
    ( ping
    , getQueryRaw
    , getQuery
    , postQueryRaw
    , postQuery
    , write
    ) where

import Database.Influx.Types
import Database.Influx.Internal.Helpers

import Data.Aeson ((.:))
import Control.Exception (catch, throw, toException, SomeException)
import Control.Monad (void)
import Data.Either (lefts, rights)
import Network.HTTP.Client.Conduit
import Network.HTTP.Simple
import Network.HTTP.Types (Status(..))
import Network.HTTP.Types.Header (ResponseHeaders)
import qualified Data.Aeson as A
import qualified Data.ByteString as BS
import qualified Data.ByteString.Lazy as LBS
import qualified Data.Conduit.Attoparsec as C
import qualified Data.Conduit.List as C
import qualified Data.Text as T
import qualified Data.Text.Encoding as T

ping :: Config -> IO (Maybe InfluxVersion)
ping config =
    do let url = configServer config `urlAppend` "/ping"
       request <- setRequestMethod "HEAD" <$> parseUrl url
       response <- httpLBS request
       let version = getResponseHeader "X-Influxdb-Version" response
       return $
           if null version || getResponseStatusCode response /= 204
             then Nothing
             else Just . InfluxVersion . T.decodeUtf8 $ head version

{-
queryRequestMethod :: Query -> Maybe BS.ByteString
queryRequestMethod q =
    case (queryStatementType q) of
      "select" ->
           if containsIntoClause q
             then Just "POST"
             else Just "GET"
      "show" -> Just "GET"
      "alter" -> Just "POST"
      "create" -> Just "POST"
      "delete" -> Just "POST"
      "drop" -> Just "POST"
      "kill" -> Just "POST"
      "grant" -> Just "POST"
       _ -> Nothing
    where
      queryStatementType =
          T.toLower . T.takeWhile (/= ' ') . T.strip . unQuery

containsIntoClause :: Query -> Bool
containsIntoClause query =
    if T.isInfixOf "into" q && T.isInfixOf "from" q
      then let tokens = T.words q
               intoIndex = "into" `elemIndex` tokens
               fromIndex = "from" `elemIndex` tokens
               selectIndex = "select" `elemIndex` tokens
          in selectIndex < intoIndex && intoIndex < fromIndex
      else False -- can't contain an INTO-clause
    where q = T.toLower . T.strip $ unQuery query

-}

queryRaw ::
       BS.ByteString -- ^ HTTP method
    -> Config
    -> QueryParams
    -> Query
    -> IO (QueryResponse [InfluxResult])
queryRaw method config params query =
    do let url = configServer config `urlAppend` "/query"
           queryString =
             maybe [] credsToQueryString (configCreds config) ++
             queryParamsToQueryString params ++
             [ ("q", Just (T.encodeUtf8 (unQuery query))) ]
       baseReq <- parseUrl url
       let req =
             setRequestMethod method $
             maybe id setRequestManager (configManager config) $
             setRequestQueryString queryString baseReq
       httpSink req responseSink `catch` \e ->
           pure $ QueryFailed $ QueryFailureHttpException e
    where
      responseSink res =
          case statusCode (responseStatus res) of
            sci | sci <= 200 && sci < 300 ->
                do parsingRes <- C.sinkParserEither A.json'
                   case parsingRes of
                     Left err -> pure $ QueryFailed $ QueryResponseJsonParseError $ show err
                     Right val ->
                         case A.fromJSON val of
                           A.Error err -> pure $ QueryFailed $ QueryResponseJsonParseError err
                           A.Success (InfluxResults results) -> pure $ QueryResult results
            400 ->
                do bs <- LBS.fromChunks <$> C.consume
                   pure $ QueryFailed $ BadInfluxQueryRequest $ responseErrorMsg (bs <$ res)
            sci -> fail $ "unexpected status code: " ++ show sci

getQueryRaw :: Config -> QueryParams -> Query -> IO (QueryResponse [InfluxResult])
getQueryRaw = queryRaw "GET"

postQueryRaw :: Config -> QueryParams -> Query -> IO (QueryResponse [InfluxResult])
postQueryRaw = queryRaw "POST"

postQuery :: Config -> Maybe DatabaseName -> Query -> IO ()
postQuery config mDatabase query =
    let params = defaultQueryParams { qp_database = mDatabase }
    in void (postQueryRaw config params query)

parseInfluxTable ::
  FromInfluxPoint t
  => InfluxTable
  -> ParsedTable t
parseInfluxTable table =
    let parseIfPossible row =
            case parseEither parseInfluxPoint Nothing row of
              Left _err -> Left row
              Right parsed -> Right parsed
        xs = map parseIfPossible (tableValues table)
        parsedRows = rights xs
        pointsThatCouldNotBeParsed = lefts xs
    in ParsedTable {..}

getQuery ::
    FromInfluxPoint t
    => Config
    -> Maybe DatabaseName
    -> Query
    -> IO (ParsedTable t)
getQuery config mDatabase query =
    do let params = defaultQueryParams { qp_database = mDatabase }
       queryRes <- getQueryRaw config params query
       case queryRes of
         QueryResult results -> parseResults results
         QueryFailed err -> throw err
    where
      parseResults results =
         case results of
           [] -> fail "no result"
           _:_:_ -> fail "multiple results"
           [result] ->
               case resultTables result of
                 Nothing -> fail "result has no points!"
                 Just tables ->
                     case tables of
                       [] -> fail "no tables"
                       _:_:_ -> fail "multiple tables"
                       [table] -> pure (parseInfluxTable table)

newtype JsonErrorResponse
    = JsonErrorResponse
    { _unJsonErrorResponse :: T.Text
    } deriving (Show)

instance A.FromJSON JsonErrorResponse where
    parseJSON =
        A.withObject "JsonErrorResponse" $ \o ->
            JsonErrorResponse <$> o .: "error"

-- | Try to parse the response as JSON of the format { "error": "some error string" } to get the
-- error message. If this fails, just return the whole response body.
responseErrorMsg :: Response LBS.ByteString -> T.Text
responseErrorMsg res =
    case A.decode (responseBody res) of
      Just (JsonErrorResponse t) -> t
      Nothing -> T.decodeUtf8 $ LBS.toStrict $ responseBody res

checkStatusWithExpectedErrorCodes ::
       [Int]
    -> Status
    -> ResponseHeaders
    -> CookieJar
    -> Maybe SomeException
checkStatusWithExpectedErrorCodes expectedErrorCodes status@(Status sci _) resHeaders cookieJar =
    if (200 <= sci && sci < 300) || sci `elem` expectedErrorCodes
      then Nothing
      else
          Just $ toException $
              StatusCodeException status resHeaders cookieJar

write ::
       Config
    -> WriteParams
    -> DatabaseName
    -> [InfluxData]
    -> IO WriteResponse
write config params database ds =
    do let url = configServer config `urlAppend` "/write"
           queryString =
             [ ("db", Just (T.encodeUtf8 database)) ] ++
             maybe [] credsToQueryString (configCreds config) ++
             writeParamsToQueryString params
           reqBody =
             RequestBodyBS $ T.encodeUtf8 $ T.unlines $
             map serializeInfluxData ds
       baseReq <- parseUrl url
       let req =
             (setRequestMethod "POST" $
             setQueryString queryString $
             maybe id setRequestManager (configManager config) $
             setRequestBody reqBody baseReq)
             { checkStatus = checkStatusWithExpectedErrorCodes [400, 404, 500] }
       (httpLBS req >>= handleResponse) `catch` \e ->
           pure $ WriteFailed $ WriteFailureHttpException e
    where
      handleResponse res =
          case statusCode (responseStatus res) of
            sci | sci <= 200 && sci < 300 ->
                pure WriteSuccessful
            400 ->
                pure $ WriteFailed $ BadInfluxWriteRequest $ responseErrorMsg res
            404 ->
                pure $ WriteFailed $ InfluxDbDoesNotExist $ responseErrorMsg res
            500 ->
                pure $ WriteFailed $ InfluxServerError $ responseErrorMsg res
            sci -> fail $ "unexpected status code: " ++ show sci
